### ISSUE TITLE
Add NFNets to keras.applications

### ISSUE THREAD

**Issue by marcusturewicz** [tensorflow/tensorflow#47233]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): 2.4.1
- Are you willing to contribute it (Yes/No): Yes, as much as possible but probably need help.

**Describe the feature and the current behavior/state.**

DeepMind recently introduced their new "Normalizer-Free" [NFNets](https://arxiv.org/abs/2102.06171), matching the test accuracy of EfficientNet-B7 on ImageNet while being up to 8.7x faster to train, and their largest models attain a new state-of-the-art top-1 accuracy of 86.5%. It seems fitting that these would be available in keras since they surpass EfficientNets which are already available.

**Will this change the current api? How?**

This would introduce new APIs i.e., `tensorflow.keras.applications.NFNetF0` through to `tensorflow.keras.applications.NFNetF6`.

**Who will benefit with this feature?**

Anyone doing image classification, including transfer learning and fine tuning; a large portion of the machine learning community!

**Any Other info.**

DeepMind's [implementation](https://github.com/deepmind/deepmind-research/tree/master/nfnets)

**Comment by jeromepatel** [comment]
I would like to work on this issue. If contributors can guide me through the solution approach, it would be helpful

**Comment by ysyyork** [comment]
I can provide implementation for WSConv and AGCModel by the way

**Comment by raulcarlomagno** [comment]
what about this?
https://github.com/ypeleg/nfnets-keras
maybe some code could be reused

**Comment by ysyyork** [comment]
> what about this?
> https://github.com/ypeleg/nfnets-keras
> maybe some code could be reused

I referred to this but it's not working out of box. it's converted from the original pytorch code without a careful transition to keras. Let me just copy paste my code here for whoever wanna work on this:

```
import tensorflow as tf
from tensorflow.python.keras.engine import data_adapter

from tensorflow.python.ops import array_ops, nn_ops, nn
from tensorflow.keras.layers import *
import tensorflow.keras.backend as K


def _standardize_weight(weight, eps, dtype, gain=None):
    mean = tf.math.reduce_mean(weight, axis=(0, 1, 2), keepdims=True)
    var = tf.math.reduce_variance(weight, axis=(0, 1, 2), keepdims=True)
    fan_in = tf.cast(tf.math.reduce_prod(weight.shape[:-1]), dtype)

    updated_weight = (weight - mean) / tf.math.rsqrt((var * fan_in + tf.convert_to_tensor(eps, dtype=dtype)))
    if gain is not None:
        updated_weight = updated_weight * gain
    return updated_weight

class WSConv2D(Conv2D):
    def __init__(self, *args,
                 eps=1e-4,
                 enable_gain=False,
                 **kwargs):
        super(WSConv2D, self).__init__(*args, **kwargs)
        self.eps = eps
        self.enable_gain = enable_gain

    def build(self, input_shape):
        if self.enable_gain:
            self.gain = self.add_weight(
                name='gain',
                shape=(1, 1, 1, self.filters),
                initializer='ones',
                trainable=True,
                dtype=self.dtype
            )
        return super().build(input_shape)

    def call(self, inputs):
        weight = self.kernel
        if self.enable_gain:
            updated_kernel = _standardize_weight(weight, self.eps, self.dtype, self.gain)
        else:
            updated_kernel = _standardize_weight(weight, self.eps, self.dtype)

        if self._is_causal:  # Apply causal padding to inputs for Conv1D.
            inputs = array_ops.pad(inputs, self._compute_causal_padding(inputs))

        outputs = self._convolution_op(inputs, updated_kernel)

        if self.use_bias:
            output_rank = outputs.shape.rank
            if self.rank == 1 and self._channels_first:
                # nn.bias_add does not accept a 1D input tensor.
                bias = array_ops.reshape(self.bias, (1, self.filters, 1))
                outputs += bias
            else:
                # Handle multiple batch dimensions.
                if output_rank is not None and output_rank > 2 + self.rank:

                    def _apply_fn(o):
                        return nn.bias_add(o, self.bias, data_format=self._tf_data_format)

                    outputs = nn_ops.squeeze_batch_dims(
                        outputs, _apply_fn, inner_rank=self.rank + 1)
                else:
                    outputs = nn.bias_add(
                        outputs, self.bias, data_format=self._tf_data_format)

        if self.activation is not None:
            return self.activation(outputs)
        return outputs

    def get_config(self):
        orig_config = super().get_config()
        orig_config['eps'] = self.eps
        return orig_config

class AGCModel(tf.keras.Model):
    def __init__(self, *args, clip_factor=0.01, eps=1e-3, **kwargs):
        super(AGCModel, self).__init__(*args, **kwargs)
        self.clip_factor = clip_factor
        self.eps = eps

    @staticmethod
    def compute_norm(x, axis, keepdims):
        return tf.math.reduce_sum(x ** 2, axis=axis, keepdims=keepdims) ** 0.5

    @staticmethod
    def unitwise_norm(x):
        if len(x.get_shape()) <= 1:  # Scalars and vectors
            axis = None
            keepdims = False
        elif len(x.get_shape()) in [2, 3]:  # Linear layers of shape IO or multihead linear
            axis = 0
            keepdims = True
        elif len(x.get_shape()) == 4:  # Conv kernels of shape HWIO
            axis = [0, 1, 2, ]
            keepdims = True
        else:
            raise ValueError(f"Got a parameter with shape not in [1, 2, 4]! {x}")
        return AGCModel.compute_norm(x, axis, keepdims)

    def adaptive_clip_grad(self, parameters, gradients):
        new_grads = []
        for (params, grads) in zip(parameters, gradients):
            p_norm = self.unitwise_norm(params)
            max_norm = tf.math.maximum(p_norm, self.eps) * self.clip_factor
            grad_norm = self.unitwise_norm(grads)
            clipped_grad = grads * (max_norm / tf.math.maximum(grad_norm, 1e-6))
            new_grad = tf.where(grad_norm < max_norm, grads, clipped_grad)
            new_grads.append(new_grad)
        return new_grads

    def train_step(self, data):
        data = data_adapter.expand_1d(data)
        x, y, sample_weight = data_adapter.unpack_x_y_sample_weight(data)

        with tf.GradientTape() as tape:
            y_pred = self(x, training=True)
            loss = self.compiled_loss(y, y_pred, sample_weight, regularization_losses=self.losses)
        trainable_params = self.trainable_variables
        gradients = tape.gradient(loss, trainable_params)
        agc_gradients = self.adaptive_clip_grad(trainable_params, gradients)
        self.optimizer.apply_gradients(zip(agc_gradients, trainable_params))
        self.compiled_metrics.update_state(y, y_pred, sample_weight)
        return {m.name: m.result() for m in self.metrics}

```

**Comment by marcusturewicz** [comment]
A PyTorch implementation for reference https://github.com/rwightman/pytorch-image-models/blob/master/timm/models/nfnet.py

**Comment by innat** [comment]
Another PyTorch implementation for ref. https://github.com/vballoli/nfnets-pytorch

**Comment by Sicily-F** [comment]
has there been any progress on this? I'd love to use on Tensorflow/Keras too!!

**Comment by batuhaneralp** [comment]
Are there any news about it yet? I see the post from 12 days ago but is this discussed within?

**Comment by innat** [comment]
Any update on this?

**Comment by sumansahoo16** [comment]
Is it happening?

**Comment by Tungway1990** [comment]
May I know the progress? Do you need manpower to help on this feature update?

**Comment by Saduf2019** [comment]
@marcusturewicz 

Please post this issue on [keras-team/keras repo.](https://github.com/keras-team/keras/issues)
To know more refer to:
https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999

**Comment by marcusturewicz** [comment]
I have transferred this to https://github.com/keras-team/keras/issues/15229, so closing this one.
